API_DESIGN.md

```diff
@@ -0,0 +1,217 @@
+Use domain-specific terminology and avoid abbreviations.
+For example, use `OrganizationID` instead of **OrgID** or **resourceOwner** for the creation of a new user or when returning one.
+
+> [!NOTE]
+> We'll update the resources in the [concepts section](https://zitadel.com/docs/concepts/structure/instance) to describe
+> common resources and their meaning.
+> Until then, please refer to the following issue: https://github.com/zitadel/zitadel/issues/5888
+
+#### Resources and Fields
+
+When a context is required for creating a resource, the context is added as a field to the resource.
+For example, when creating a new user, the organization ID is required. The `organization_id` is added as a field to the `CreateUserRequest`.
+
+Only allow providing a context where it is required. Do not provide the possibility to provide a context where it is not required.
+For example, when retrieving or updating a user, the organization ID is not required, since the user can be determined by the user ID.
+However, it is possible to provide the organization ID as a filter to retrieve a list of users of a specific organization.
+
+Prevent the creation of global messages that are used in multiple resources unless they always follow the same pattern.
+Use dedicated fields as described above or create a separate message for the specific context, that is only used in the boundary of the same resource.  
+For example, settings might be set as a default on the instance level, but might be overridden on the organization level.
+In this case, the settings could share the same `SettingsContext` message to determine the context of the settings.
+But do not create a global `Context` message that is used across the whole API if there are different scenarios and different fields required for the context.  
+The same applies to messages that are returned by multiple resources.  
+For example, information about the `User` might be different when managing the user resource itself than when it's returned
+as part of an authorization or a manager role, where only limited information is needed.
+
+Prevent reusing messages for the creation and the retrieval of a resource.
+Returning messages might contain additional information that is not required or even not available for the creation of the resource.  
+What might sound obvious when designing the CreateUserRequest for example, where only an `organization_id` but not the 
+`organization_name` is available, might not be so obvious when designing some sub-resource like a user's `IdentityProviderLink`, 
+which might contain an `identity_provider_name` when returned but not when created.
+
+#### Operations and Methods
+
+Methods on a resource should be named using the following convention:
+- Create: `Create<resource>`
+- Update: `Update<resource>`
+- Delete: `Delete<resource>`
+- Get: `Get<resource>`
+- List: `List<resource>`
+- Search: `Search<resource>`
```

Comment:
A description would help, questions i have:
- what happens if the resource already exists on create
- is update partial
- what happens if resource does not exist on delete?
- whats the difference between list and search?

Reply:
swithced to a table and removed `Search`